### DISCUSSION TITLE
OSGi "Test-Bundle-Specification"

### DISCUSSION THREAD

**Discussion by laeubi** [osgi/osgi#547] [Ideas]
When talking to others, the most unclear part for people about OSGi often is:

> What is the preferred way to test OSGi code, is there an official recommendation?

Of course there are multiple ways and different levels and some approaches, but all of them lack one important part: There is no common specification / definition so I can't exchange them and there is no common base even many of them share similar concepts.

Here are some examples:
- BND-Testing: https://bnd.bndtools.org/chapters/310-testing.html
- Tycho-Test-Plugin-Packaging: https://tycho.eclipseprojects.io/doc/latest/tycho-surefire-plugin/test-mojo.html
- Pax Exam-Container: https://ops4j1.jira.com/wiki/spaces/PAXEXAM4/overview

All of them has (in some way or the other) some way to identify test, fire up a framework, and then use a "test-provider" to execute the test-cases, but of course I can't just share one bundle used for bnd with one from pax or tycho or ...

I therefore think it would be good to have a very basic "Test-Bundle-Specification" that tools can leverage to give users a guidance how they should "shape" their test-bundles so they can be executed by compatible tools.

As a first start, I think these minimal definitions would give a good starting point for a first spec-draft that might be build upon then:

- Defining a way how to identify a bundle as containing Test classes, BND uses `Test-Cases` header what seems suitable and flexible enough for a generic definition
- Defining a way how to execute these tests, Tycho uses "test-providers" (similar to maven-surefire), a more modern approach would be to define some kind of `Test-Provider-Capability`

If that's done, it would be great to also define how a compliant test execution could/should work e.g.
- should there be a `TestExecutorService` where one can trigger tests?
- should tests automatically executed once such a "Test-bundle" is deployed and resolved
- how should results be published (e.g. result file, vendor specific, DTO model ...)

**Comment by bjhargrave** [comment]
> What is the preferred way to test OSGi code, is there an official recommendation?

I think you can look to the osgi-test project for an OSGi recommended way to test OSGi code. If fact we use osgi-test in the OSGi TCKs.

**Reply by juergen-albert** [reply]
Adding something like a  unified Testcases Header/Capability to the Manifest wouldn't be a bad idea. With that Test bundles become the chance to be useful across build tools.

**Reply by bjhargrave** [reply]
I am not sure what you mean. Is this not the Test-Cases header?

**Reply by laeubi** [reply]
osgi-test is an implementation not a specification ... `Test-Cases` header is not osgi-test but bnd-test-specific extension (even though it does not follow OSGI-best-practise adding an `X-` prefix)

**Reply by juergen-albert** [reply]
`Test-Cases` is a Header with meaning to bnd. I propose (and believe @laeubi as well) to add something like this as e.g. a specified Header in one of our Specifications. We could e.g. add it to the list here: https://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#i2654895

**Reply by bjhargrave** [reply]
> to add something like this as e.g. a specified Header in one of our Specifications

This is not something the specs should have. The OSGi specs are for runtime. We should not spec how we build and test things. We can certainly document (not in a spec) how we build and test things.

**Reply by juergen-albert** [reply]
| The OSGi specs are for runtime.

Mostly yes, but our Annotations describe what a build tool should do with it. It thus describes something, that has no influence at runtime. I also don't see where we describe what influence e.g. the specified Header `Bundle-Vendor: ` has at runtime.

**Reply by laeubi** [reply]
> > to add something like this as e.g. a specified Header in one of our Specifications
> 
> This is not something the specs should have. The OSGi specs are for runtime. We should not spec how we build and test things. We can certainly document (not in a spec) how we build and test things.

Well in the last spec call is was argued that a TCK must "run in a real OSGi Framework" so actually the tests are executed at runtime... But how do I "really run" them if running tests is nowhere specified?

So for me this is nothing one can really separate as not runtime-related...

**Reply by bjhargrave** [reply]
> Mostly yes, but our Annotations describe what a build tool should do with it. It thus describes something, that has no influence at runtime.

That is not true. Of course the annotation cause the tool to generate information which is then used at runtime.

**Reply by laeubi** [reply]
I'm not sure if arguing about the definitions of "runtime" helps, there are actually sometime "concepts" described in the spec e.g.: 

- https://docs.osgi.org/specification/osgi.core/8.0.0/framework.module.html#framework.module-semantic.versioning and
- https://docs.osgi.org/specification/osgi.core/8.0.0/framework.module.html#framework.module-semantic.versioning.type.roles

that are "not enforced at runtime" and therefore, strictly speaking, not relevant for the runtime.

And as described at the bottom of my idea, of course one could think about having a running OSGi-Framework and then deploy a "Test-Bundle" and having some Agent execute it and report back results... one even can think about having the TCKs deployed and having such a service check the impls first before allow the bundle to be seen by others so if one just want to enforce some "runtime aspect" I'm sure one can find some :-)

E.g. the [Eclipse-Test-Framework](https://wiki.eclipse.org/Platform-releng/Eclipse_Test_Framework) works that way that the is a Eclipse-Osgi-Framework running and you deploy test-bundles into it that runs acceptance test for different platforms without recompile the code or even have a certain build tool!

**Comment by juergen-albert** [comment]
I would vote raising this to an issue and to discuss this in the next spec call.

**Reply by laeubi** [reply]
@juergen-albert I think discussions can be discussed in the spec call as well, but you probably want to add a label to it (I think only collaborators of the repository can add labels).

**Reply by juergen-albert** [reply]
not sure what kind of lable I should give it.

**Reply by laeubi** [reply]
`spec-call` ? if later on we came up with a better name you can just rename labels as well.

**Comment by laeubi** [comment]
For example https://bnd.bndtools.org/chapters/310-testing.html already has some kind of "specification" of how test are run in a runtime see section "Testing With JUnit 5 - biz.aQute.tester.junit-platform":

> This tester will:
> 
>   -  scan all running bundles for any JUnit Platform-compliant TestEngine implementations;
>   -  scan all running bundles for any bundles with the Test-Cases header set
>  -   execute all of the discovered tests for each of the discovered engines.
